Can't hook app into Red Hat managed apicurio service

### Describe the bug

I'm trying to "hook" my application into the Red Hat managed service registry/kafka by following [this blog post](https://developers.redhat.com/articles/2022/03/23/quarkus-superheroes-managed-services-save-day).

The Quarkus superheroes app runs fine against the apicurio instance that gets deployed if you run `oc apply -f https://raw.githubusercontent.com/quarkusio/quarkus-super-heroes/main/deploy/k8s/java17-openshift.yml`, but once trying to follow the guide and connect to the managed services, the app blows up with the following exception.

[The blog post](https://developers.redhat.com/articles/2022/03/23/quarkus-superheroes-managed-services-save-day) was written with Quarkus `2.7.3`. I tried with Quarkus `2.7.5` and its still broken, with each Quarkus version up to and including `2.8.2`.

The app does seem to successfully create the schema within the managed service registry instance.

```
20:22:34 ERROR [io.sm.re.me.kafka] (smallrye-kafka-producer-thread-0) SRMSG18260: Unable to recover from the serialization failure (topic: fights), configure a SerializationFailureHandler to recover from errors.: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "references" (class io.apicurio.registry.rest.v2.beans.ArtifactMetaData), not marked as ignorable (15 known properties: "modifiedBy", "createdOn", "modifiedOn", "name", "state", "properties", "contentId", "version", "type", "id", "description", "globalId", "createdBy", "groupId", "labels"])
 at [Source: (byte[])"{"name":"Fight","createdBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","createdOn":"2022-05-04T20:22:34+0000","modifiedBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","modifiedOn":"2022-05-04T20:22:34+0000","id":"fights-value","version":"1","type":"AVRO","globalId":1,"state":"ENABLED","contentId":1,"references":[]}"; line: 1, column: 358] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["references"])
	at io.apicurio.rest.client.util.ConcurrentUtil.get(ConcurrentUtil.java:29)
	at io.apicurio.rest.client.util.ConcurrentUtil.get(ConcurrentUtil.java:13)
	at io.apicurio.rest.client.util.ConcurrentUtil.result(ConcurrentUtil.java:40)
	at io.apicurio.rest.client.VertxHttpClient.sendRequest(VertxHttpClient.java:122)
	at io.apicurio.registry.rest.client.impl.RegistryClientImpl.createArtifact(RegistryClientImpl.java:236)
	at io.apicurio.registry.rest.client.RegistryClient.createArtifact(RegistryClient.java:139)
	at io.apicurio.registry.resolver.DefaultSchemaResolver.lambda$handleAutoCreateArtifact$2(DefaultSchemaResolver.java:189)
	at io.apicurio.registry.resolver.ERCache.lambda$getValue$0(ERCache.java:132)
	at io.apicurio.registry.resolver.ERCache.retry(ERCache.java:171)
	at io.apicurio.registry.resolver.ERCache.getValue(ERCache.java:131)
	at io.apicurio.registry.resolver.ERCache.getByContent(ERCache.java:116)
	at io.apicurio.registry.resolver.DefaultSchemaResolver.handleAutoCreateArtifact(DefaultSchemaResolver.java:187)
	at io.apicurio.registry.resolver.DefaultSchemaResolver.resolveSchema(DefaultSchemaResolver.java:90)
	at io.apicurio.registry.serde.AbstractKafkaSerializer.serialize(AbstractKafkaSerializer.java:83)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.lambda$serialize$1(SerializerWrapper.java:56)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.wrapSerialize(SerializerWrapper.java:81)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.serialize(SerializerWrapper.java:56)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:945)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:905)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaProducer.lambda$send$3(ReactiveKafkaProducer.java:111)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.ExecutionException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "references" (class io.apicurio.registry.rest.v2.beans.ArtifactMetaData), not marked as ignorable (15 known properties: "modifiedBy", "createdOn", "modifiedOn", "name", "state", "properties", "contentId", "version", "type", "id", "description", "globalId", "createdBy", "groupId", "labels"])
 at [Source: (byte[])"{"name":"Fight","createdBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","createdOn":"2022-05-04T20:22:34+0000","modifiedBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","modifiedOn":"2022-05-04T20:22:34+0000","id":"fights-value","version":"1","type":"AVRO","globalId":1,"state":"ENABLED","contentId":1,"references":[]}"; line: 1, column: 358] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["references"])
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.apicurio.rest.client.util.ConcurrentUtil.get(ConcurrentUtil.java:21)
	... 26 more
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "references" (class io.apicurio.registry.rest.v2.beans.ArtifactMetaData), not marked as ignorable (15 known properties: "modifiedBy", "createdOn", "modifiedOn", "name", "state", "properties", "contentId", "version", "type", "id", "description", "globalId", "createdBy", "groupId", "labels"])
 at [Source: (byte[])"{"name":"Fight","createdBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","createdOn":"2022-05-04T20:22:34+0000","modifiedBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","modifiedOn":"2022-05-04T20:22:34+0000","id":"fights-value","version":"1","type":"AVRO","globalId":1,"state":"ENABLED","contentId":1,"references":[]}"; line: 1, column: 358] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["references"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:61)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownProperty(DeserializationContext.java:1127)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:1989)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1700)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1678)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:319)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:176)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4674)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3707)
	at io.apicurio.rest.client.response.ResponseHandler.handle(ResponseHandler.java:50)
	at io.apicurio.rest.client.response.ResponseHandler.handle(ResponseHandler.java:18)
	at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:400)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:387)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:365)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:332)
	at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:294)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$null$8(HttpContext.java:550)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63)
	at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```

Follow the steps in https://developers.redhat.com/articles/2022/03/23/quarkus-superheroes-managed-services-save-day

Troubleshooting a bit further...

I even went back to the version I used when I wrote that article and played with the `rest-fights` app & I see the same error.

I've put together some smaller yamls that can be imported via `oc apply -f` which only contain the `rest-fights` service showing the issue. I'm deploying on [Red Hat Developer Sandbox](https://developers.redhat.com/developer-sandbox/get-started), as described in the blog post.

[rest-fights-java17.zip](https://github.com/quarkusio/quarkus/files/8631508/rest-fights-java17.zip)

If you unzip this there is a single folder `rest-fights-java17`. Inside are 2 yaml files: `rest-fights-quarkus-2.7.3-java17.yml` and `rest-fights-quarkus-2.8.2-java17.yml`.
- `rest-fights-quarkus-2.7.3-java17.yml` is for Quarkus `2.7.3`, which is the version I used when I created [the blog post](https://developers.redhat.com/articles/2022/03/23/quarkus-superheroes-managed-services-save-day), so I know this worked!
- `rest-fights-quarkus-2.8.2-java17.yml` is for Quarkus `2.8.2`, the most recent version of the app. They both suffer from this problem.

Before deploying either you first need to [create the managed services as described in the blog post](https://developers.redhat.com/articles/2022/03/23/quarkus-superheroes-managed-services-save-day?source=sso#creating_managed_services).

> **NOTE:** For this simple example you **ONLY** need to create the `rest-fights` service account & permissions on the services.

Then before deploying the app (via `oc apply -f`) you need to update a few values in the respective yaml file:
- In the `rest-fights-config` ConfigMap (1st element in the yaml document) you need to update the values:
    - `kafka.bootstrap.servers`: Update `<kafka-bootstrap-server>` with the bootstrap server of your Kafka managed service
    - `mp.messaging.connector.smallrye-kafka.apicurio.registry.url`: Update `<service-registry-core-registry-api>` with the core registry API URL of your service registry service
- In the `rest-fights-config-creds` Secret (2nd element in the yaml document) you need to update the values:
    - `CLIENT_ID`: Update `<rest-fights-client-id>` with the Client ID of the service account you created
    - `CLIENT_SECRET`: Update `<rest-fights-client-secret>` with the Client secret of the service account you created

Then `oc apply -f` the one you want to deploy and wait for everything to start up (in my instructions below I'm assuming the `2.7.3` version - the only difference with the `2.8.2` version, besides dependency updates, is swapping in MongoDB instead of PostgreSQL.

![image](https://user-images.githubusercontent.com/363447/166929345-59fe7f8c-88cf-47d6-a441-ba3893e49e48.png)

Everything should start up fine. Then click the route icon on the `ui-super-heroes` app to bring up the UI

![Screen Shot 2022-05-05 at 9 08 59 AM](https://user-images.githubusercontent.com/363447/166929690-d7fb77be-9b28-4781-bd1f-d6ba32fbcbc1.png)

You should see fallback hero & fallback villain (because we did not deploy the rest of the system). Just click `Fight!`

![Screen Shot 2022-05-05 at 9 10 21 AM](https://user-images.githubusercontent.com/363447/166929905-9768c4f5-fdad-411d-a59c-3f0b69a6661e.png)

Then go and look at the pod logs for the `rest-fights` pod:

```
ERROR [io.sm.re.me.kafka] (smallrye-kafka-producer-thread-0) SRMSG18260: Unable to recover from the serialization failure (topic: fights), configure a SerializationFailureHandler to recover from errors.: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "references" (class io.apicurio.registry.rest.v2.beans.ArtifactMetaData), not marked as ignorable (15 known properties: "modifiedBy", "createdOn", "modifiedOn", "name", "state", "properties", "contentId", "version", "type", "id", "description", "globalId", "createdBy", "groupId", "labels"])
 at [Source: (byte[])"{"name":"Fight","createdBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","createdOn":"2022-05-04T20:22:34+0000","modifiedBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","modifiedOn":"2022-05-04T20:22:34+0000","id":"fights-value","version":"1","type":"AVRO","globalId":1,"state":"ENABLED","contentId":1,"references":[]}"; line: 1, column: 358] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["references"])
	at io.apicurio.rest.client.util.ConcurrentUtil.get(ConcurrentUtil.java:29)
	at io.apicurio.rest.client.util.ConcurrentUtil.get(ConcurrentUtil.java:13)
	at io.apicurio.rest.client.util.ConcurrentUtil.result(ConcurrentUtil.java:40)
	at io.apicurio.rest.client.VertxHttpClient.sendRequest(VertxHttpClient.java:122)
	at io.apicurio.registry.rest.client.impl.RegistryClientImpl.createArtifact(RegistryClientImpl.java:236)
	at io.apicurio.registry.rest.client.RegistryClient.createArtifact(RegistryClient.java:139)
	at io.apicurio.registry.resolver.DefaultSchemaResolver.lambda$handleAutoCreateArtifact$2(DefaultSchemaResolver.java:189)
	at io.apicurio.registry.resolver.ERCache.lambda$getValue$0(ERCache.java:132)
	at io.apicurio.registry.resolver.ERCache.retry(ERCache.java:171)
	at io.apicurio.registry.resolver.ERCache.getValue(ERCache.java:131)
	at io.apicurio.registry.resolver.ERCache.getByContent(ERCache.java:116)
	at io.apicurio.registry.resolver.DefaultSchemaResolver.handleAutoCreateArtifact(DefaultSchemaResolver.java:187)
	at io.apicurio.registry.resolver.DefaultSchemaResolver.resolveSchema(DefaultSchemaResolver.java:90)
	at io.apicurio.registry.serde.AbstractKafkaSerializer.serialize(AbstractKafkaSerializer.java:83)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.lambda$serialize$1(SerializerWrapper.java:56)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.wrapSerialize(SerializerWrapper.java:81)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.serialize(SerializerWrapper.java:56)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:945)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:905)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaProducer.lambda$send$3(ReactiveKafkaProducer.java:111)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.ExecutionException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "references" (class io.apicurio.registry.rest.v2.beans.ArtifactMetaData), not marked as ignorable (15 known properties: "modifiedBy", "createdOn", "modifiedOn", "name", "state", "properties", "contentId", "version", "type", "id", "description", "globalId", "createdBy", "groupId", "labels"])
 at [Source: (byte[])"{"name":"Fight","createdBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","createdOn":"2022-05-04T20:22:34+0000","modifiedBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","modifiedOn":"2022-05-04T20:22:34+0000","id":"fights-value","version":"1","type":"AVRO","globalId":1,"state":"ENABLED","contentId":1,"references":[]}"; line: 1, column: 358] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["references"])
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.apicurio.rest.client.util.ConcurrentUtil.get(ConcurrentUtil.java:21)
	... 26 more
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "references" (class io.apicurio.registry.rest.v2.beans.ArtifactMetaData), not marked as ignorable (15 known properties: "modifiedBy", "createdOn", "modifiedOn", "name", "state", "properties", "contentId", "version", "type", "id", "description", "globalId", "createdBy", "groupId", "labels"])
 at [Source: (byte[])"{"name":"Fight","createdBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","createdOn":"2022-05-04T20:22:34+0000","modifiedBy":"service-account-srvc-acct-10e65eb4-bbe3-4e65-ab1f-9e336da98c2b","modifiedOn":"2022-05-04T20:22:34+0000","id":"fights-value","version":"1","type":"AVRO","globalId":1,"state":"ENABLED","contentId":1,"references":[]}"; line: 1, column: 358] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["references"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:61)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownProperty(DeserializationContext.java:1127)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:1989)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1700)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1678)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:319)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:176)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4674)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3707)
	at io.apicurio.rest.client.response.ResponseHandler.handle(ResponseHandler.java:50)
	at io.apicurio.rest.client.response.ResponseHandler.handle(ResponseHandler.java:18)
	at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:400)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:387)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:365)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:332)
	at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:294)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$null$8(HttpContext.java:550)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63)
	at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -191,7 +191,7 @@
         <log4j-jboss-logmanager.version>1.3.0.Final</log4j-jboss-logmanager.version>
         <avro.version>1.11.0</avro.version>
         <apicurio-registry.version>2.2.4.Final</apicurio-registry.version>
-        <apicurio-common-rest-client.version>0.1.9.Final</apicurio-common-rest-client.version> <!-- must be the version Apicurio Registry uses -->
+        <apicurio-common-rest-client.version>0.1.11.Final</apicurio-common-rest-client.version> <!-- must be the version Apicurio Registry uses -->
         <jacoco.version>0.8.8</jacoco.version>
         <testcontainers.version>1.17.2</testcontainers.version> <!-- Make sure to also update docker-java.version to match its needs -->
         <docker-java.version>3.2.13</docker-java.version> <!-- must be the version Testcontainers use -->

